File: android-native/app/src/main/kotlin/com/sabq/smart/data/LikesStore.kt
package com.sabq.smart.data

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.sabq.smart.data.api.SabqApi
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.update

private val Context.likesDataStore by preferencesDataStore(name = "likes_prefs")
private val LIKES_KEY = stringSetPreferencesKey("liked_article_ids")

/**
 * Local-first cache of which articles the current user has liked.
 * Mirrors iOS [LikesStore] (Stores/LikesStore.swift) 1:1.
 */
@Singleton
class LikesStore @Inject constructor(
    @ApplicationContext private val context: Context,
    private val api: SabqApi,
) {
    /** Reactive flow of the currently-liked article IDs. */
    val likedIds: Flow<Set<String>> = context.likesDataStore.data
        .map { prefs -> prefs[LIKES_KEY] ?: emptySet() }

    /**
     * Last-known server count per article, keyed by id. The heart
     * reads liked-or-not from likedIds (instant), the number next to
     * it reads from this map (refreshed in the background).
     */
    private val _counts = MutableStateFlow<Map<String, Int>>(emptyMap())
    val counts: StateFlow<Map<String, Int>> = _counts.asStateFlow()

    suspend fun currentLikedIds(): Set<String> = likedIds.first()

    fun isLiked(articleId: String): Flow<Boolean> = likedIds.map { it.contains(articleId) }

    fun likesCount(articleId: String): Flow<Int?> = counts.map { it[articleId] }

    /**
     * Optimistic toggle: flip local immediately, then sync to server.
     * On server success we trust the response. On failure we revert.
     */
    suspend fun toggle(articleId: String): Pair<Boolean, Int>? {
        val wasLiked = currentLikedIds().contains(articleId)

        // Optimistically update local DataStore
        context.likesDataStore.edit { prefs ->
            val existing = prefs[LIKES_KEY] ?: emptySet()
            prefs[LIKES_KEY] = if (wasLiked) existing - articleId else existing + articleId
        }

        return try {
            val response = api.toggleArticleLike(articleId)
            val liked = response.liked
            val count = response.likesCount

            // Sync with backend truth
            context.likesDataStore.edit { prefs ->
                val existing = prefs[LIKES_KEY] ?: emptySet()
                prefs[LIKES_KEY] = if (liked) existing + articleId else existing - articleId
            }
            _counts.update { it + (articleId to count) }
            Pair(liked, count)
        } catch (e: Exception) {
            // Revert optimistic update
            context.likesDataStore.edit { prefs ->
                val existing = prefs[LIKES_KEY] ?: emptySet()
                prefs[LIKES_KEY] = if (wasLiked) existing + articleId else existing - articleId
            }
            null
        }
    }

    /**
     * On article view appear: fetch the server count + reconcile.
     * If local says liked but server says not, we POST again to heal
     * the missing row. If local says NOT liked but server says liked,
     * we just sync local to match.
     */
    suspend fun reconcile(articleId: String) {
        try {
            val response = api.fetchArticleLikeStatus(articleId)
            val liked = response.liked
            val count = response.likesCount

            _counts.update { it + (articleId to count) }

            val localSaysLiked = currentLikedIds().contains(articleId)
            if (liked && !localSaysLiked) {
                context.likesDataStore.edit { prefs ->
                    val existing = prefs[LIKES_KEY] ?: emptySet()
                    prefs[LIKES_KEY] = existing + articleId
                }
            } else if (!liked && localSaysLiked) {
                // Heal — push the like back to the server so the row exists
                runCatching { api.toggleArticleLike(articleId) }
            }
        } catch (e: Exception) {
            // Network blip — keep local state as-is
        }
    }

    suspend fun clear() {
        _counts.value = emptyMap()
        context.likesDataStore.edit { it.remove(LIKES_KEY) }
    }
}

File: sabq app ios/sabq/Stores/LikesStore.swift
import SwiftUI

// Local-first cache of which articles the current user has liked.
// Modeled on BookmarksStore — addresses the editor's complaint
// 2026-05-19 evening: "بمجرد الخروج والرجوع الدخول ألقى تسجيلي للايك
// اختفى". Old flow drove `isLiked` purely off the server response,
// so any transient server hiccup (token expiry mid-session, a
// silently-failed insert, a 5xx) made the heart go cold on re-entry.
// Bookmarks didn't have the same bug because they kept a UserDefaults
// set as their source of truth and treated the API call as
// fire-and-forget; this store does the same for likes.
//
// Cross-device sync nuance: if the user un-likes on web while iOS is
// offline, iOS will still show the article as liked until the next
// foreground reconcile lands. The reconcile path (toggle that hits
// server "already liked, deleting" before iOS knows) self-heals on
// the next tap. That window is small enough to accept; the
// alternative — trusting the server unconditionally — is the bug we
// just fixed.
@Observable
final class LikesStore {
    private(set) var likedIDs: Set<String> = []
    /// Last-known server count per article, keyed by id. The heart
    /// reads liked-or-not from likedIDs (instant), the number next to
    /// it reads from this dict (refreshed in the background).
    private(set) var counts: [String: Int] = [:]

    private let storageKey = "sabq_liked_articles_v1"

    init() {
        if let data = UserDefaults.standard.data(forKey: storageKey),
           let decoded = try? JSONDecoder().decode(Set<String>.self, from: data) {
            likedIDs = decoded
        }
    }

    func isLiked(_ articleID: String) -> Bool { likedIDs.contains(articleID) }
    func count(for articleID: String) -> Int? { counts[articleID] }

    /// Optimistic toggle: flip local immediately, then sync to server.
    /// On server success we trust the response (which is the same
    /// state we flipped to, except in the rare case the server
    /// already had the opposite state — in which case we sync to it).
    /// On failure we revert.
    @MainActor
    func toggle(_ articleID: String) async -> (liked: Bool, count: Int)? {
        let wasLiked = likedIDs.contains(articleID)
        if wasLiked {
            likedIDs.remove(articleID)
        } else {
            likedIDs.insert(articleID)
        }
        persist()

        do {
            let result = try await BehaviorTracker.shared.toggleLike(articleId: articleID)
            if result.liked {
                likedIDs.insert(articleID)
            } else {
                likedIDs.remove(articleID)
            }
            counts[articleID] = result.count
            persist()
            return (result.liked, result.count)
        } catch {
            // Revert optimistic update.
            if wasLiked {
                likedIDs.insert(articleID)
            } else {
                likedIDs.remove(articleID)
            }
            persist()
            return nil
        }
    }

    /// On article view appear: fetch the server count + reconcile.
    /// If local says liked but server says not, we POST again to heal
    /// the missing row — this is the specific fallback that fixes the
    /// "like vanishes on re-entry" report. If local says NOT liked
    /// but server says liked (e.g. liked from web), we just sync
    /// local to match.
    @MainActor
    func reconcile(_ articleID: String) async {
        do {
            let status = try await BehaviorTracker.shared.fetchLikeStatus(articleId: articleID)
            counts[articleID] = status.count

            let localSaysLiked = likedIDs.contains(articleID)
            if status.liked && !localSaysLiked {
                likedIDs.insert(articleID)
                persist()
            } else if !status.liked && localSaysLiked {
                // Heal — push the like back to the server so the row
                // exists on next reconcile. We don't flip local
                // because the user's intent on this device was
                // "liked".
                _ = try? await BehaviorTracker.shared.toggleLike(articleId: articleID)
            }
        } catch {
            // Network blip — keep local state as-is.
        }
    }

    func clear() {
        likedIDs.removeAll()
        counts.removeAll()
        UserDefaults.standard.removeObject(forKey: storageKey)
    }

    private func persist() {
        if let data = try? JSONEncoder().encode(likedIDs) {
            UserDefaults.standard.set(data, forKey: storageKey)
        }
    }
}
